### DISCUSSION TITLE
TMDL format with object level granularity

### DISCUSSION THREAD

**Discussion by AlexisOlson** [TabularEditor/TabularEditor3#1198] [Ideas]
### Describe the solution you'd like

I love the folder format where each table, column, measure, etc. is represented as an individual file but I also love the TMDL format that doesn't have all the quotes and brackets that the JSON format has.

Any chance we can get the best of both worlds?

### Goals

1. Each object like a measure should be represented as an individual file rather than lumped together at the table level.
2. Each individual file should resemble the TMDL format rather than the JSON format.

**Comment by otykier** [comment]
Hi Alexis
Nothing would make me happier, but at the moment, the TMDL serializer provided by Microsoft, does not have a lot of configuration options, so unfortunately there’s no way to configure which objects should get serialized as individual files. Creating our own serializer would be a relatively large undertaking, so it’s not something we can prioritize at the moment. However, I’ve given the same feedback to Microsoft several times that this is needed. If more people voice their opinion on this, I’m sure they’ll listen!

**Reply by AlexisOlson** [reply]
That makes sense. Thanks for clarifying.

If you submit the idea to [ideas.powerbi.com](https://ideas.fabric.microsoft.com/), I'll see if I can gather support. :)
(I could submit the idea, but I think you'd be better at precisely articulating the request.)

**Comment by otykier** [comment]
An update on this: We don't expect Microsoft to provide file-level customization for the TMDL serializer. However, Microsoft has provided an API that lets us serialize individual TOM objects, meaning we could potentially build this type of functionality on our side.

We would like to hear how the Tabular Editor community feels about this? Is this something that's urgently needed? Also, what kind of customization would you prefer, if you were able to store objects as individual TMDL files?

**Reply by corycundy** [reply]
Daniel,

I think the ability to serialize at the measure level is the most important. As measures change over time, it make it super easy to see the history of a measure within a git repo. To me separating tables and columns is more of a nice to have. Column properties change much less significantly and I almost think it may be easier to view history on an entire table object. This is especially true if the column order is deterministic within a table file and not changing around when minor updates are made. The bim file struggles with this greatly across the different object types.

In summary, I do think it is important to separate at least measures, but my guess is that if you go through the effort to separate measure, columns aren't a far stretch.

Lastly, while I think this is important, I still vote for Calculation Item debugging as a much higher priority and secondly Query Scoped measure debugging. These features would really fill the gaps in the debugger!

Cory Cundy

**Reply by AlexisOlson** [reply]
I'm not planning to use TMDL until it can support measures as separate objects, so it's pretty important if TMDL is important.

**Reply by adamharperCrisp** [reply]
individual object serialization (like the folder structure) allows for much less git conflicts

agree measures are the most important

**Comment by ScottThompsonNH** [comment]
Yes, object serialization, particularly for measures would be great.

**Comment by AlexisOlson** [comment]
I asked program manager Ruy Romano about this on LinkedIn. [He replied](https://www.linkedin.com/feed/update/urn:li:activity:7165748464666984448?commentUrn=urn%3Ali%3Acomment%3A%28activity%3A7165748464666984448%2C7170961916410912768%29&replyUrn=urn%3Ali%3Acomment%3A%28activity%3A7165748464666984448%2C7171128096987725824%29&dashCommentUrn=urn%3Ali%3Afsd_comment%3A%287170961916410912768%2Curn%3Ali%3Aactivity%3A7165748464666984448%29&dashReplyUrn=urn%3Ali%3Afsd_comment%3A%287171128096987725824%2Curn%3Ali%3Aactivity%3A7165748464666984448%29) that it's

> On the backlog. If you dont mind, please create a new idea in [fabric.ideas.com](http://fabric.ideas.com/)

**Reply by AlexisOlson** [reply]
I've posted this on the ideas site.

Please vote for it here:
https://ideas.fabric.microsoft.com/ideas/idea/?ideaid=a72123c6-b2fd-ee11-a73c-0022485327bf
